### DISCUSSION TITLE
Cannot create executable publisher and subscriber files

### DISCUSSION THREAD

**Discussion by MuhammedErkam** [OpenDDS/OpenDDS#3804] [Q&A] [answered]
Hi,
I am new to openDDS and trying to understand Introduction to OpenDDS example at this site: https://opendds.org/about/articles/Article-Intro.html

Then, 
I followed all instructions and wrote codes and produced files but when I am typing "$ACE_ROOT/bin/mwc.pl -type gnuace StockQuoter.mwc" instruction to build project but I cannot have publisher and subscriber executables.

So, I cannot run project.
For example when I try to run subcriber "./subscriber -DCPSConfigFile dds_tcp_conf.ini"
I get this message  "bash: ./subscriber: No such file or directory"

**Comment by iguessthislldo** [comment]
You need to run `make` after running `mwc.pl` if you haven't already. That is what will build the the applications. The article doesn't really say it explicitly, so I can see how it could confuse someone. `mwc.pl` is just what generates the files that make (or Visual Studio, if on Windows) uses to build a project.

**Reply by MuhammedErkam** [reply]
I also tried that one but at this time I get below errors. I did not change anything in the code and did not understand how to solve errors.


make[1]: Entering directory '/home/erkam/src/OpenDDS/examples/DCPS/denemestock'

GNUmakefile: /home/erkam/src/OpenDDS/examples/DCPS/denemestock/GNUmakefile.StockQuoter_Common MAKEFLAGS=w

make[1]: Leaving directory '/home/erkam/src/OpenDDS/examples/DCPS/denemestock'
make[1]: Entering directory '/home/erkam/src/OpenDDS/examples/DCPS/denemestock'

GNUmakefile: /home/erkam/src/OpenDDS/examples/DCPS/denemestock/GNUmakefile.StockQuoter_Publisher MAKEFLAGS=w

g++ -fvisibility=hidden -fvisibility-inlines-hidden -Wnon-virtual-dtor -ggdb -pthread -fno-strict-aliasing -Wall -W -Wpointer-arith -pipe -D_GNU_SOURCE   -I/home/erkam/src/OpenDDS/ACE_wrappers -D__ACE_INLINE__ -I/home/erkam/src/OpenDDS/ACE_wrappers -I/home/erkam/src/OpenDDS/ACE_wrappers/TAO -I../../.. -I/home/erkam/src/OpenDDS/ACE_wrappers/TAO/orbsvcs  -Wl,-E -L/home/erkam/src/OpenDDS/ACE_wrappers/lib -L/home/erkam/src/OpenDDS/lib -L.  -L/home/erkam/src/OpenDDS/ACE_wrappers/lib  -L../../../lib   -o publisher .obj/publisher.o -lStockQuoterCommon -lTAO_Svc_Utils -lOpenDDS_Tcp -lOpenDDS_InfoRepoDiscovery -lOpenDDS_Dcps -lTAO_BiDirGIOP -lTAO_PI -lTAO_CodecFactory -lTAO_PortableServer -lTAO_AnyTypeCode -lTAO -lACE -ldl -lrt 
.obj/publisher.o: In function `main':
/home/erkam/src/OpenDDS/examples/DCPS/denemestock/publisher.cpp:132: undefined reference to `StockQuoter::QuoteDataWriter::_narrow(CORBA::Object*)'
/home/erkam/src/OpenDDS/examples/DCPS/denemestock/publisher.cpp:149: undefined reference to `StockQuoter::ExchangeEventDataWriter::_narrow(CORBA::Object*)'
.obj/publisher.o: In function `StockQuoter::QuoteTypeSupportImpl::QuoteTypeSupportImpl()':
/home/erkam/src/OpenDDS/examples/DCPS/denemestock/StockQuoterTypeSupportImpl.h:193: undefined reference to `VTT for StockQuoter::QuoteTypeSupportImpl'
/home/erkam/src/OpenDDS/examples/DCPS/denemestock/StockQuoterTypeSupportImpl.h:193: undefined reference to `VTT for StockQuoter::QuoteTypeSupportImpl'
/home/erkam/src/OpenDDS/examples/DCPS/denemestock/StockQuoterTypeSupportImpl.h:193: undefined reference to `VTT for StockQuoter::QuoteTypeSupportImpl'
/home/erkam/src/OpenDDS/examples/DCPS/denemestock/StockQuoterTypeSupportImpl.h:193: undefined reference to `StockQuoter::QuoteTypeSupport::QuoteTypeSupport()'
/home/erkam/src/OpenDDS/examples/DCPS/denemestock/StockQuoterTypeSupportImpl.h:193: undefined reference to `VTT for StockQuoter::QuoteTypeSupportImpl'
/home/erkam/src/OpenDDS/examples/DCPS/denemestock/StockQuoterTypeSupportImpl.h:193: undefined reference to `VTT for StockQuoter::QuoteTypeSupportImpl'
/home/erkam/src/OpenDDS/examples/DCPS/denemestock/StockQuoterTypeSupportImpl.h:193: undefined reference to `VTT for StockQuoter::QuoteTypeSupportImpl'
/home/erkam/src/OpenDDS/examples/DCPS/denemestock/StockQuoterTypeSupportImpl.h:193: undefined reference to `VTT for StockQuoter::QuoteTypeSupportImpl'
/home/erkam/src/OpenDDS/examples/DCPS/denemestock/StockQuoterTypeSupportImpl.h:193: undefined reference to `VTT for StockQuoter::QuoteTypeSupportImpl'
/home/erkam/src/OpenDDS/examples/DCPS/denemestock/StockQuoterTypeSupportImpl.h:193: undefined reference to `vtable for StockQuoter::QuoteTypeSupportImpl'
/home/erkam/src/OpenDDS/examples/DCPS/denemestock/StockQuoterTypeSupportImpl.h:193: undefined reference to `vtable for StockQuoter::QuoteTypeSupportImpl'
/home/erkam/src/OpenDDS/examples/DCPS/denemestock/StockQuoterTypeSupportImpl.h:193: undefined reference to `vtable for StockQuoter::QuoteTypeSupportImpl'
/home/erkam/src/OpenDDS/examples/DCPS/denemestock/StockQuoterTypeSupportImpl.h:193: undefined reference to `vtable for StockQuoter::QuoteTypeSupportImpl'
/home/erkam/src/OpenDDS/examples/DCPS/denemestock/StockQuoterTypeSupportImpl.h:193: undefined reference to `vtable for StockQuoter::QuoteTypeSupportImpl'
.obj/publisher.o:/home/erkam/src/OpenDDS/examples/DCPS/denemestock/StockQuoterTypeSupportImpl.h:193: more undefined references to `vtable for StockQuoter::QuoteTypeSupportImpl' follow
.obj/publisher.o: In function `StockQuoter::QuoteTypeSupportImpl::QuoteTypeSupportImpl()':
/home/erkam/src/OpenDDS/examples/DCPS/denemestock/StockQuoterTypeSupportImpl.h:193: undefined reference to `VTT for StockQuoter::QuoteTypeSupportImpl'
/home/erkam/src/OpenDDS/examples/DCPS/denemestock/StockQuoterTypeSupportImpl.h:193: undefined reference to `VTT for StockQuoter::QuoteTypeSupportImpl'
/home/erkam/src/OpenDDS/examples/DCPS/denemestock/StockQuoterTypeSupportImpl.h:193: undefined reference to `StockQuoter::QuoteTypeSupport::~QuoteTypeSupport()'
/home/erkam/src/OpenDDS/examples/DCPS/denemestock/StockQuoterTypeSupportImpl.h:193: undefined reference to `VTT for StockQuoter::QuoteTypeSupportImpl'
/home/erkam/src/OpenDDS/examples/DCPS/denemestock/StockQuoterTypeSupportImpl.h:193: undefined reference to `VTT for StockQuoter::QuoteTypeSupportImpl'
.obj/publisher.o: In function `StockQuoter::ExchangeEventTypeSupportImpl::ExchangeEventTypeSupportImpl()':
/home/erkam/src/OpenDDS/examples/DCPS/denemestock/StockQuoterTypeSupportImpl.h:428: undefined reference to `VTT for StockQuoter::ExchangeEventTypeSupportImpl'
/home/erkam/src/OpenDDS/examples/DCPS/denemestock/StockQuoterTypeSupportImpl.h:428: undefined reference to `VTT for StockQuoter::ExchangeEventTypeSupportImpl'
/home/erkam/src/OpenDDS/examples/DCPS/denemestock/StockQuoterTypeSupportImpl.h:428: undefined reference to `VTT for StockQuoter::ExchangeEventTypeSupportImpl'
/home/erkam/src/OpenDDS/examples/DCPS/denemestock/StockQuoterTypeSupportImpl.h:428: undefined reference to `StockQuoter::ExchangeEventTypeSupport::ExchangeEventTypeSupport()'
/home/erkam/src/OpenDDS/examples/DCPS/denemestock/StockQuoterTypeSupportImpl.h:428: undefined reference to `VTT for StockQuoter::ExchangeEventTypeSupportImpl'
/home/erkam/src/OpenDDS/examples/DCPS/denemestock/StockQuoterTypeSupportImpl.h:428: undefined reference to `VTT for StockQuoter::ExchangeEventTypeSupportImpl'
/home/erkam/src/OpenDDS/examples/DCPS/denemestock/StockQuoterTypeSupportImpl.h:428: undefined reference to `VTT for StockQuoter::ExchangeEventTypeSupportImpl'
/home/erkam/src/OpenDDS/examples/DCPS/denemestock/StockQuoterTypeSupportImpl.h:428: undefined reference to `VTT for StockQuoter::ExchangeEventTypeSupportImpl'
/home/erkam/src/OpenDDS/examples/DCPS/denemestock/StockQuoterTypeSupportImpl.h:428: undefined reference to `VTT for StockQuoter::ExchangeEventTypeSupportImpl'
/home/erkam/src/OpenDDS/examples/DCPS/denemestock/StockQuoterTypeSupportImpl.h:428: undefined reference to `vtable for StockQuoter::ExchangeEventTypeSupportImpl'
/home/erkam/src/OpenDDS/examples/DCPS/denemestock/StockQuoterTypeSupportImpl.h:428: undefined reference to `vtable for StockQuoter::ExchangeEventTypeSupportImpl'
/home/erkam/src/OpenDDS/examples/DCPS/denemestock/StockQuoterTypeSupportImpl.h:428: undefined reference to `vtable for StockQuoter::ExchangeEventTypeSupportImpl'
/home/erkam/src/OpenDDS/examples/DCPS/denemestock/StockQuoterTypeSupportImpl.h:428: undefined reference to `vtable for StockQuoter::ExchangeEventTypeSupportImpl'
/home/erkam/src/OpenDDS/examples/DCPS/denemestock/StockQuoterTypeSupportImpl.h:428: undefined reference to `vtable for StockQuoter::ExchangeEventTypeSupportImpl'
.obj/publisher.o:/home/erkam/src/OpenDDS/examples/DCPS/denemestock/StockQuoterTypeSupportImpl.h:428: more undefined references to `vtable for StockQuoter::ExchangeEventTypeSupportImpl' follow
.obj/publisher.o: In function `StockQuoter::ExchangeEventTypeSupportImpl::ExchangeEventTypeSupportImpl()':
/home/erkam/src/OpenDDS/examples/DCPS/denemestock/StockQuoterTypeSupportImpl.h:428: undefined reference to `VTT for StockQuoter::ExchangeEventTypeSupportImpl'
/home/erkam/src/OpenDDS/examples/DCPS/denemestock/StockQuoterTypeSupportImpl.h:428: undefined reference to `VTT for StockQuoter::ExchangeEventTypeSupportImpl'
/home/erkam/src/OpenDDS/examples/DCPS/denemestock/StockQuoterTypeSupportImpl.h:428: undefined reference to `StockQuoter::ExchangeEventTypeSupport::~ExchangeEventTypeSupport()'
/home/erkam/src/OpenDDS/examples/DCPS/denemestock/StockQuoterTypeSupportImpl.h:428: undefined reference to `VTT for StockQuoter::ExchangeEventTypeSupportImpl'
/home/erkam/src/OpenDDS/examples/DCPS/denemestock/StockQuoterTypeSupportImpl.h:428: undefined reference to `VTT for StockQuoter::ExchangeEventTypeSupportImpl'
.obj/publisher.o: In function `TAO_Objref_Var_T<StockQuoter::QuoteTypeSupport>::~TAO_Objref_Var_T()':
/home/erkam/src/OpenDDS/ACE_wrappers/TAO/tao/Objref_VarOut_T.inl:18: undefined reference to `TAO::Objref_Traits<StockQuoter::QuoteTypeSupport>::release(StockQuoter::QuoteTypeSupport*)'
.obj/publisher.o: In function `TAO_Objref_Var_T<StockQuoter::ExchangeEventTypeSupport>::~TAO_Objref_Var_T()':
/home/erkam/src/OpenDDS/ACE_wrappers/TAO/tao/Objref_VarOut_T.inl:18: undefined reference to `TAO::Objref_Traits<StockQuoter::ExchangeEventTypeSupport>::release(StockQuoter::ExchangeEventTypeSupport*)'
.obj/publisher.o: In function `TAO_Objref_Var_T<StockQuoter::QuoteDataWriter>::~TAO_Objref_Var_T()':
/home/erkam/src/OpenDDS/ACE_wrappers/TAO/tao/Objref_VarOut_T.inl:18: undefined reference to `TAO::Objref_Traits<StockQuoter::QuoteDataWriter>::release(StockQuoter::QuoteDataWriter*)'
.obj/publisher.o: In function `TAO_Objref_Var_T<StockQuoter::ExchangeEventDataWriter>::~TAO_Objref_Var_T()':
/home/erkam/src/OpenDDS/ACE_wrappers/TAO/tao/Objref_VarOut_T.inl:18: undefined reference to `TAO::Objref_Traits<StockQuoter::ExchangeEventDataWriter>::release(StockQuoter::ExchangeEventDataWriter*)'
collect2: error: ld returned 1 exit status
GNUmakefile.StockQuoter_Publisher:185: recipe for target 'publisher' failed
make[1]: *** [publisher] Error 1
make[1]: Leaving directory '/home/erkam/src/OpenDDS/examples/DCPS/denemestock'
GNUmakefile:45: recipe for target 'StockQuoter_Publisher' failed
make: *** [StockQuoter_Publisher] Error 2

**Comment by mitza-oci** [comment, marked as answer]
The code for this article is included in the OpenDDS repository in `examples/DCPS/IntroductionToOpenDDS`

**Reply by MuhammedErkam** [reply]
Yes, I know that but it is working already. I copied publisher.cpp, subscriber.cpp, StockQouter.idl, .mwc and .mpc files to new folder. Then I ran below commands in order. Then make gave me above errors. I couldn't solve them.

$DDS_ROOT/bin/opendds_idl StockQuoter.idl
tao_idl --unknown-annotations ignore --idl-version 4 -I$DDS_ROOT -I$TAO_ROOT/orbsvcs StockQuoter.idl 
tao_idl --unknown-annotations ignore --idl-version 4 -I$DDS_ROOT -I$TAO_ROOT/orbsvcs StockQuoterTypeSupport.idl 
$ACE_ROOT/bin/mwc.pl -type gnuace StockQuoter.mwc
make

**Reply by mitza-oci** [reply]
This shouldn't need manual invocation of the code generators, the rules to run those are built in to the makefiles.
Does building within the OpenDDS source tree work?  If so, what's different between that build and the one you're setting up?

**Reply by MuhammedErkam** [reply]
I realized that I do not have to do these commands:

`$DDS_ROOT/bin/opendds_idl StockQuoter.idl
tao_idl --unknown-annotations ignore --idl-version 4 -I$DDS_ROOT -I$TAO_ROOT/orbsvcs StockQuoter.idl
tao_idl --unknown-annotations ignore --idl-version 4 -I$DDS_ROOT -I$TAO_ROOT/orbsvcs StockQuoterTypeSupport.idl`


But when I do this commands in order:

`$ACE_ROOT/bin/mwc.pl -type gnuace StockQuoter.mwc`
`make`

I get this error:

> make[1]: Entering directory '/home/erkam/src/OpenDDS/examples/DCPS/xxx'
> 
> GNUmakefile: /home/erkam/src/OpenDDS/examples/DCPS/xxx/GNUmakefile.StockQuoter_Common MAKEFLAGS=w
> 
> g++ -fvisibility=hidden -fvisibility-inlines-hidden -Wnon-virtual-dtor -ggdb -pthread -fno-strict-aliasing -Wall -W -Wpointer-arith -pipe -D_GNU_SOURCE   -I/home/erkam/src/OpenDDS/ACE_wrappers -D__ACE_INLINE__ -I/home/erkam/src/OpenDDS/ACE_wrappers -I/home/erkam/src/OpenDDS/ACE_wrappers/TAO -I../../.. -I/home/erkam/src/OpenDDS/ACE_wrappers/TAO/orbsvcs -DSTOCKQUOTERCOMMON_BUILD_DLL  -c -fPIC -o .shobj/StockQuoterTypeSupportC.o StockQuoterTypeSupportC.cpp
> In file included from StockQuoterTypeSupportC.cpp:32:0:
> StockQuoterTypeSupportC.h:44:15: fatal error: StockQuoterCommon_Export.h: No such file or directory
>  #include /**/ "StockQuoterCommon_Export.h"
>                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
> compilation terminated.
> /home/erkam/src/OpenDDS/ACE_wrappers/include/makeinclude/rules.local.GNU:189: recipe for target '.shobj/StockQuoterTypeSupportC.o' failed
> make[1]: *** [.shobj/StockQuoterTypeSupportC.o] Error 1
> make[1]: Leaving directory '/home/erkam/src/OpenDDS/examples/DCPS/xxx'
> GNUmakefile:37: recipe for target 'StockQuoter_Common' failed
> make: *** [StockQuoter_Common] Error 2

StockQuoterCommon_Export.h cannot produced here.

**Reply by iguessthislldo** [reply]
Yes, `StockQuoterCommon_Export.h` isn't made as part of the build, [it's already part of the source code](https://github.com/objectcomputing/OpenDDS/blob/master/examples/DCPS/IntroductionToOpenDDS/StockQuoterCommon_Export.h). Copy it from `examples/DCPS/IntroductionToOpenDDS/StockQuoterCommon_Export.h` into your project.

**Reply by MuhammedErkam** [reply]
Thank you so much for helping me :)
